### PR TITLE
Creates a new Collector Level Type

### PR DESCRIPTION
**Pull request by Hamms**
**Notes**
- [Spec here](https://docs.google.com/document/d/14DftAnydxozz4bEmo7JuOjKE7SwCCgtXt2Wm8jrB7Zw/edit#)
- Depends on #9224  
- Note that the assets are currently those we use for artist, and should be regarded as temporary placeholder. Furthermore, this feature should be regarded as not ready for public consumption until we obtain new art assets.

**Description**

Collector Levels are a new Maze level type intended to emphasize the importance of block limits while providing a flexible number of in-level goals. Students will be successful if they are able to collect _any number_ of the available collectibles within the block limits. Using too many blocks or collecting nothing will be a a failure, and managing to collect all collectibles will be considered "perfect".

**Screenshots**

| pic | description |
| --- | --- |
| image | Level with sample art |
| image | new blocks |
| image | exceeded block limit |
| image | collected none |
| image | collected some |
| image | collected all |

### CODE REVIEW

File: apps/src/maze/collectorBlocks.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+exports.install = function (blockly, blockInstallOptions) {
+  var generator = blockly.Generator.get('JavaScript');
+  blockly.JavaScript = generator;
+
+  // Block for collecting collectibles
+  blockUtils.generateSimpleBlock(blockly, generator, {
+    name: 'collector_collect',
+    helpUrl: 'TODO',
+    title: 'collect',
+    tooltip: 'TODO',
```
Comment (joshlory) [inline apps/src/maze/collectorBlocks.js]: TODO.

File: apps/src/maze/collectorBlocks.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      this.appendStatementInput('DO')
+          .appendTitle(msg.doCode());
+      this.setTooltip(msg.ifTooltip());
+      this.setPreviousStatement(true);
+      this.setNextStatement(true);
+    }
+  };
+
+  generator.collector_ifCollectible = function () {
+    var argument = 'Maze.pilePresent(\'block_id_' + this.id + '\')';
```
Comment (joshlory) [inline apps/src/maze/collectorBlocks.js]: These are good candidates for string interpolation:

``` javascript
`Maze.pilePresent('block_id_${this.id}')`
```

File: apps/src/maze/collectorDrawer.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+ * Extends Drawer to draw collectibles for Collector
+ */
+export default class CollectorDrawer extends Drawer {
+
+  /**
+   * @override
+   */
+  updateImageWithIndex_(prefix, row, col) {
+    let img = super.updateImageWithIndex_(prefix, row, col);
+    let val = this.map_.getValue(row, col);
```
Comment (joshlory) [inline apps/src/maze/collectorDrawer.js]: Nit: I'm not completely sure how these are consumed, but can some of the `let` definitions in this file be `const`?

(Brent has suggested defaulting to `const` everywhere, unless the value needs to change — still camelCase.)

File: apps/src/maze/maze.js [hunk 4]
```diff
@@ -1850,6 +1868,18 @@ function isDirtCorrect() {
 }
 
 /**
+ * Certain Maze types - namely, WordSearch, Collector, and any Maze with
+ * Quantum maps, don't want to check for success until the user's conde
```
Comment (joshlory) [inline apps/src/maze/maze.js]: Nit: typo ("user's code").

File: apps/src/maze/mazeMap.js [hunk 5]
```diff
@@ -24,16 +24,23 @@ MazeMap.prototype.forEachCell = function (cb) {
   });
 };
 
+MazeMap.prototype.getCell = function (x, y) {
+  return this.currentStaticGrid[x] && this.currentStaticGrid[x][y];
```
Comment (joshlory) [inline apps/src/maze/mazeMap.js:28]: I could very well be doing something wrong, but I get an exception here if I delete 'Value' but forget to clear 'Range':

screen shot 2016-07-06 at 10 51 13 pm
`Cannot read property '0' of undefined`

**Comment by joshlory** [comment]
Hmm, it's been a while since I've built blockly-core.  Even after `./build_with_core.sh debug` I get `blockly.js:14950 Uncaught Error: "collector_collect" is an unknown language block.`

**Comment by Hamms** [comment]
That's odd. In what context is that error?

File: apps/src/maze/collectorBlocks.js (hunk 2 above)
Comment (Hamms) [inline apps/src/maze/collectorBlocks.js]: oh man, _so many_ of the custom blocks could benefit from string interpolation.

File: apps/src/maze/mazeMap.js (hunk 5 above)
Comment (Hamms) [inline apps/src/maze/mazeMap.js:28]: Hmmm. I do not. You might need to build code-studio.

Comment (Hamms) [inline apps/src/maze/mazeMap.js:28]: I cannot repro this. Could I get more details of your repro steps?

**Comment by joshlory** [comment]
Followed up offline.  LGTM!

screen shot 2016-07-07 at 3 57 03 pm

**Pull request merged**